File: b191-p215-Multithreaded.py
#!/bin/env python3.6

# proof-of-work-example.py
# now multi-threaded!

import hashlib
import time
import threading

max_nonce = (2 ** 31) - 1
THREADS = 1


# multi-thread memory space
# each thread has its own entry in the array
class Result:
    res = []

    def __init__(self):
        for t in range(THREADS):
            self.res.append(0)


def worker(header, nonce, target, thread_nr, memory_pool):
    hash_result = hashlib.sha256(str(str(header)+str(nonce)).encode("UTF-8")).hexdigest()

    if int(hash_result, 16) < target:
        print("Sucsess with nonce %d" % nonce)
        print("Hash is %s" % hash_result)
        memory_pool.res[thread_nr] = nonce
        return
    memory_pool.res[thread_nr] = -1


def proof_of_work(header, difficulty_bits, result):
    target = 2 ** (256-difficulty_bits)

    for nonce in range(0, max_nonce, THREADS):
        threads = []
        for i in range(THREADS):
            t = threading.Thread(target=worker, name="worker" + str(i + 1), args=(header, nonce + i, target, i, result))
            threads.append(t)
            t.start()

        success = False
        success_thread = -1
        for i in range(THREADS):
            threads[i].join()
            if result.res[i] is not -1:
                success = True
                success_thread = i

        if success:
            return nonce + success_thread


    print("Failed after %d (max_nonce) tries" % nonce)
    return nonce


if __name__ == "__main__":
    nonce       = 0
    hash_result = ""
    result = Result()

    for difficulty_bits in range(32):
        difficulty = 2 ** difficulty_bits
        print("Difficulty: %ld (%d bits)" % (difficulty, difficulty_bits))

        print("Starting search...")

        start_time = time.time()

        new_block = "test block with transactions" + hash_result

        nonce = proof_of_work(new_block, difficulty_bits, result)
        hash_result = str(nonce)

        end_time = time.time()

        elapsed_time = end_time - start_time
        print("Elapsed time: %.4f seconds" % elapsed_time)

        if elapsed_time > 0:
            hash_power = float(int(nonce)/elapsed_time)
            print("Hashing Power: %ld hashes per second" % hash_power)

        print("")

File: b191-p215-multithreaded.cpp


//// imports ////////////////////////////////////
// primitives
#include <string>	// text
#include <vector>	// vector
#include <iostream>	// print
#include <stdint.h>	// formated int

// threading
#include <atomic>	// primitives for multithread acsess
#include <thread>	// threading

// ssl
#include <openssl/sha.h> // crypto



//// Constants //////////////////////////////////
uint8_t CORES;
if (thread::hardware_concurrency() > 1){
	const CORES = thread::hardware_concurrency() - 1;
}
else {
	const CORES = 1;
}



//// datatypes //////////////////////////////////

// contains the results from the workers
struct result {
	std::vector<atomic_ullong, CORES> result;
	std::vector<atomic_ullong, CORES> nonce;
	std::vector<atomic_ullong, CORES> iterations = 0;
	std::atomic_short finnished = -2;
} ;

// contains previus block hash and next hash salt
typedef struct {
	std::atomic_uchar prevHash[32];
	uint64_t nonce = 0;
} __attribute__((packed)) block;



//// functions //////////////////////////////////



//// Main thread /////////////////////////////////
char proof_of_woork(uint64_t max_nonce, block working_block) {
	double difficulty;
	double difficulty_bits;
	double target;
	result res;

	std::vector<std::thread, CORES> threads;
	uint64_t iterations;

	time_t start_time;
	time_t end_time;
	
	difficulty = 2;
	for (uint64_t j = 0; j < 256; j++) {
		difficulty = difficulty * difficulty;
	}

	for (uint64_t i = 0; i < 32; i++) {
		difficulty_bits = 2;
		for (uint64_t j = 0; j < i; j++) {
			difficulty_bits = difficulty_bits * difficulty_bits;
		}

		target = difficulty - difficulty_bits;
		

		iterations = 0;

		cout << "Difficulty: " << difficulty;
		cout << "Starting shearch...";

		time(&start_time);

		// create threads
		for (uint64_t j = 0; j < CORES; j++) {
			std::thread thr(worker, working_block, j, res, difficulty, max_nonce);
			threads[j] = thr;
		}

		// wait for threads to finnish
		for (uint64_t j = 0; j < threads.size(); j++) {
			threads[j].join();
		}

		time(&end_time);

		// extract results
		for (uint64_t j = 0; j < CORES; j++) {
			iterations += res.iterations;
		}

		if (res.finnished > -1) {
			working_block.prevHash = res.result[res.finnished];
			working_block.nonce = res.nonce[res.finnished];
			cout << "Sucsess! hash found at nonce: " << to_string(working_block.nonce);
			cout << "Hash is: " << to_string(working_block.prevHash);
			cout << "Hash per seond is: " << to_string(working_block.nonce / difftime(end_time, start_time));
		}
		else if (working_block.finnished == -1) {
			cout << "found no results after " << max_nonce << "tries";
		}
		else if (working_block.finnished == -2) {
			cout << "Error: threads still running!";
		}
	}
}

// worker thread
void worker(block working_block, uint8_t thread_nr, result memory_pool, uint64_t target, uint64_t max_once) {
	char hash_result[];
	double hash_result_float;
	
	for (uint64_t i = thread_nr; i < max_once; i + CORES) {
		if (memory_pool.finnished > -2) {
			return;
		}

		working_block.nonce = i;
		memory_pool.iterations[thread_nr]++;
		
		SHA256(working_block, sizeof(block), hash_result);
		double hash_result_float = *(double*)hash_result;
		
		if (hash_result_float < target) {
			momory_pool.finnished = thread_nr;
			memory_pool.nonce[thread_nr] = nonce;
			memory_pool.result[thread_nr] = hash_result;
			release hash_result_float;
			return;
		}
	}
}



//// initialize ///////////////////////////////////////
int main() {
	cout(to_string(sizeof(block)));

	// init
	uint64_t max_once = 2 ** 63;
	
	block initial_block;
	initial_block.prevHash = "                                ";

	// compute

}
